File: backend/internal/model/oauth2_client.go
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package model

import (
	"time"

	"gorm.io/gorm"
)

// OAuth2Client represents an OIDC/OAuth2 client (application) that can use KkOps as IdP
type OAuth2Client struct {
	ID           uint           `gorm:"primaryKey" json:"id"`
	ClientID     string         `gorm:"uniqueIndex;not null;size:128" json:"client_id"`
	ClientSecret string         `gorm:"not null;size:255" json:"-"` // hashed
	Name         string         `gorm:"not null;size:100" json:"name"`
	Protocol     string         `gorm:"not null;size:16;default:oidc" json:"protocol"` // oidc | saml | ldap
	RedirectURIs string         `gorm:"type:text;not null" json:"-"`                   // JSON array of allowed redirect_uri
	Scopes       string         `gorm:"type:text" json:"-"`                            // space-separated, e.g. "openid profile email"
	Enabled      bool           `gorm:"default:true" json:"enabled"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName specifies table name
func (OAuth2Client) TableName() string {
	return "oauth2_clients"
}

File: frontend/src/api/oauth2Client.ts
// Copyright (c) 2025 kk
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

import apiClient from './client'

export interface OAuth2Client {
  id: number
  client_id: string
  client_secret?: string
  name: string
  protocol: string
  redirect_uris: string[]
  scopes: string
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface CreateOAuth2ClientRequest {
  name: string
  protocol?: string
  redirect_uris: string[]
  scopes?: string
}

export interface UpdateOAuth2ClientRequest {
  name?: string
  protocol?: string
  redirect_uris?: string[]
  scopes?: string
  enabled?: boolean
}

export const oauth2ClientApi = {
  list: (params?: { protocol?: string }) =>
    apiClient.get<{ data: OAuth2Client[] }>('/oauth2-clients', { params }),
  get: (id: number) =>
    apiClient.get<{ data: OAuth2Client }>(`/oauth2-clients/${id}`),
  create: (data: CreateOAuth2ClientRequest) =>
    apiClient.post<{ data: OAuth2Client }>('/oauth2-clients', data),
  update: (id: number, data: UpdateOAuth2ClientRequest) =>
    apiClient.put<{ data: OAuth2Client }>(`/oauth2-clients/${id}`, data),
  delete: (id: number) => apiClient.delete<unknown>(`/oauth2-clients/${id}`),
  regenerateSecret: (id: number) =>
    apiClient.post<{ data: { client_secret: string } }>(
      `/oauth2-clients/${id}/regenerate-secret`
    ),
}
